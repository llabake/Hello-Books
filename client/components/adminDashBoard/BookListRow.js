import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import swal from 'sweetalert';
import { connect } from 'react-redux';
import { Modal } from 'react-materialize';
// import { deleteBookAction } from '../../actions/bookAction';
import { deleteBookAction } from '../../actions/adminAction';
import ModifyBookDetail from './ModifyBookDetail';

/**
 * 
 * 
 * @class BookListRow
 * @extends {Component}
 */
class BookListRow extends Component {

  /**
   * Creates an instance of BookListRow.
   * @param {any} props 
   * @memberof BookListRow
   */
  constructor(props) {
    super(props);
    this.state = {
      bookToEdit : {},
      editBook: false

    }

    this.handleDelete= this.handleDelete.bind(this)
    this.handleEdit = this.handleEdit.bind(this)
  }


  /**
   * 
   * 
   * @param {any} event 
   * @memberof BookListRow
   * @returns {object} response message
   */
  handleDelete (event) {
    event.preventDefault();
    swal({
      title: 'Are you sure?',
      text: 'Once deleted, you will not be able to recover this book!',
      icon: 'warning',
      buttons: [ "Noes!", "Yes, delete it!",],
      dangerMode: true,
      closeModal: false,
    })
      .then((willDelete) => {
        if(willDelete) {
          this.props.deleteBookAction(this.props.book.id)
            .then(() => {
              swal("Poof! Your book has been deleted!", {
                icon: "success",
              });
            })
            .catch(() => {
              swal("Argh! Your book escaped delete!", {
                icon: "failure",
              });
            })
        } else {
          swal("Your book is safe!");
        }
      })
  }

  /**
   * 
   * @returns {Object} book to be edited
   * @param {any} event 
   * @memberof BookListRow
   */
  handleEdit(event) {
    event.preventDefault();
    this.setState({
      bookToEdit: this.props.book ,
      editBook: true
    }, () => {
      $('#edit-book-modal').modal('open');
    })
  }



  /**
   * 
   * 
   * @returns {object}  book object
   * @memberof BookListRow
   */
  render () {
    const { bookToEdit } = this.state
    const { book, index, } = this.props
    return (
        <tr>
          <td>{index + 1}</td>
          <td>{book.title}</td>
          <td>{book.author}</td>
          <td>
            <a onClick={this.handleEdit} 
              className="modal-trigger .modal-close"
              style={{ cursor: "pointer" }}>
            <i className="material-icons">
              edit
            </i>
            </a>
            { this.state.editBook ? 
              <Modal id='edit-book-modal'
                >
                <ModifyBookDetail book={bookToEdit} />
              </Modal >: 
              null 
            } 
          </td>
          <td>
            <a onClick={this.handleDelete}
            style={{ cursor: "pointer" }}>
            <i className="material-icons">
              delete
            </i>
            </a>
          </td>
          
        </tr>


      
    );
  }
}

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
    allBooks: state.adminReducer.allBooks,

  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    deleteBookAction: (bookId) => dispatch(deleteBookAction(bookId)),
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(BookListRow);

