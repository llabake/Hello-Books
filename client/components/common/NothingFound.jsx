import React from 'react';
import PropTypes from 'prop-types';

const NothingFound = ({ text, id }) => {
  return (
    <div
      className="card-panel row center-align"
      id={id}>
      <h4>
        {text}
      </h4>
    </div>
  )
};

NothingFound.propTypes = {
  id: PropTypes.string,
  text: PropTypes.string
};

export default NothingFound;