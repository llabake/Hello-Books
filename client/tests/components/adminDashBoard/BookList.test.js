import React from 'react';

import {
  BookList,
  mapDispatchToProps,
  mapStateToProps
} from '../../../components/adminDashBoard/BookList';

import BookListRow from '../../../components/adminDashBoard/BookListRow'
import adminData from '../../mocks/userData';
import userData from '../../mocks/userData';
import NothingFound from '../../../components/common/NothingFound';
import { Pagination } from 'react-materialize';

function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    handleShowBookList: jest.fn(),
    fetchAllBooks: jest.fn(),
    handleShowBorrowRequestList: jest.fn(),
    errors: {},
    handleSelectedPage: jest.fn(),
    loading: false,
    bookCount: 0,
    allBooks: [{
      aboutAuthor: 'Great Author',
      author: 'Imaseun Egbosa',
      borrowCount: 0, 
      createdAt: "2018-06-07T15:08:48.835Z",
      description: 'Youth juvenile deliquency',
      downVotes: 0,
      favorited: [],
      id: 5,
      image: '',
      isbn: '7698102987364',
      publishedYear: '2012',
      quantity: 7,
      reviews: [],
      title: 'Fine Boys',
      upVotes: 0,
      updatedAt: '2018-06-07T15:08:48.835Z',
      votes: []
    }],
  };

  const state = {
    maxItems: 2,
    showPagination: false,
    activePage: 1,

  };
  const wrapper = shallow(<BookList {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
  };
}

describe('The Book list tab', () => {
  test('should render and not crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should call componentDidMount', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });
    spy = jest.spyOn(BookList.prototype, 'componentDidMount')
    const { wrapper, props, state } = setup();
    expect(BookList.prototype.componentDidMount).toHaveBeenCalledTimes(1)
  })

  test('should render the book list', () => {
    const { wrapper, props, state } = setup();
    expect(wrapper.find(BookListRow).length).toBe(1);
    expect(wrapper.find(Pagination).length).toBe(0);
  })

  test('should render nothing found component', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      allBooks: [],
      bookCount: 0,
      fetchAllBooks: jest.fn(() => Promise.resolve(1)),
    }
    const wrapper = shallow(<BookList {...props} />);

    expect(wrapper.find(NothingFound).length).toBe(1)

  })

  test('should call handle selected page', () => {
    const { wrapper, props, state } = setup();
    const page = 2
    wrapper.setState({
      activePage: page,
    });
    wrapper.instance().handleSelectedPage(page)
    expect(wrapper.instance().state.activePage).toBe(2);
  });

  test('should call componentWillReceiveProps', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      allBooks: [],
      bookCount: 0,
      fetchAllBooks: jest.fn(() => Promise.resolve(1)),
    }

    const nextProps = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      allBooks: [],
      bookCount: 0,
      fetchAllBooks: jest.fn(() => Promise.resolve(1)),
    }
    spy = jest.spyOn(BookList.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<BookList {...props} />)
    wrapper.setProps(nextProps);
    expect(BookList.prototype.componentWillReceiveProps).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalled();
  })

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchAllBooks');

    const { fetchAllBooks } = mapDispatchToProps(dispatch);
    fetchAllBooks();
    expect(dispatch).toHaveBeenCalled();
  });

  test('mapStateToProps', () => {
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('allBooks');
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('loading');
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('bookCount');
  });

})