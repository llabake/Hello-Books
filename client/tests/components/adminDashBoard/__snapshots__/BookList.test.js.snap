// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Book list tab should render and not crash 1`] = `
<div
  id="allbooks"
>
  <div
    className="col s12"
  >
    <div
      className="card-panel"
    >
      <table
        className="bordered highlight responsive-table"
      >
        <thead>
          <tr>
            <th>
              S/N
            </th>
            <th>
              Title
            </th>
            <th>
              Author 
            </th>
            <th>
              Edit
            </th>
            <th>
              Delete
            </th>
          </tr>
        </thead>
        <tbody>
          <Connect(BookListRow)
            book={
              Object {
                "aboutAuthor": "Great Author",
                "author": "Imaseun Egbosa",
                "borrowCount": 0,
                "createdAt": "2018-06-07T15:08:48.835Z",
                "description": "Youth juvenile deliquency",
                "downVotes": 0,
                "favorited": Array [],
                "id": 5,
                "image": "",
                "isbn": "7698102987364",
                "publishedYear": "2012",
                "quantity": 7,
                "reviews": Array [],
                "title": "Fine Boys",
                "upVotes": 0,
                "updatedAt": "2018-06-07T15:08:48.835Z",
                "votes": Array [],
              }
            }
            index={0}
            key="5"
          />
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
